File: frameworks/motia/motia-py/playground/steps/api_steps/sse_step.py
"""Accepts URL-encoded data and streams back random items as SSE."""

import json
import math
import random
import time
from typing import Any
from urllib.parse import parse_qsl

from motia import MotiaHttpArgs, http, logger

config = {
    "name": "SSE Example",
    "description": "Accepts URL-encoded data and streams back random items as SSE",
    "flows": ["sse-example"],
    "triggers": [
        http("POST", "/sse"),
    ],
    "enqueues": [],
}


async def handler(args: MotiaHttpArgs[Any]) -> None:
    """Read URL-encoded body, then stream random items back as SSE."""
    request = args.request
    response = args.response

    safe_header_keys = ("content-type", "user-agent", "accept", "content-length", "x-request-id")
    safe_headers = {k: v for k, v in request.headers.items() if k.lower() in safe_header_keys}
    logger.info("Data received", {"headers": safe_headers})

    await response.status(200)
    await response.headers({
        "content-type": "text/event-stream",
        "cache-control": "no-cache",
        "connection": "keep-alive",
    })

    raw_chunks: list[str] = []
    for chunk in request.request_body.stream:
        if isinstance(chunk, bytes):
            raw_chunks.append(chunk.decode("utf-8", errors="replace"))
        else:
            raw_chunks.append(str(chunk))

    payload = "".join(raw_chunks).strip()
    parts: dict[str, str] = dict(parse_qsl(payload, keep_blank_values=True))

    items = _generate_random_items(parts)

    for item in items:
        response.writer.stream.write(f"event: item\ndata: {json.dumps(item)}\n\n".encode("utf-8"))
        time.sleep(0.3 + random.random() * 0.7)

    response.writer.stream.write(f"event: done\ndata: {json.dumps({'total': len(items)})}\n\n".encode("utf-8"))
    response.close()


def _generate_random_items(parts: dict[str, str]) -> list[dict[str, Any]]:
    fields = [{"name": k, "value": v} for k, v in parts.items()]
    count = 5 + math.floor(random.random() * 6)

    adjectives = ["swift", "lazy", "bold", "calm", "fierce", "gentle", "sharp", "wild"]
    nouns = ["falcon", "river", "mountain", "crystal", "thunder", "shadow", "ember", "frost"]

    return [
        {
            "id": f"item-{int(time.time() * 1000)}-{i}",
            "label": f"{random.choice(adjectives)} {random.choice(nouns)}",
            "score": round(random.random() * 100),
            "source": fields[i % len(fields)] if fields else None,
        }
        for i in range(count)
    ]

File: frameworks/motia/motia-js/playground/steps/api-steps/sse.step.ts
import { type Handlers, http, logger, type StepConfig } from 'motia'

export const config = {
  name: 'SSE Example',
  description: 'Accepts URL-encoded data and streams back random items as SSE',
  flows: ['sse-example'],
  triggers: [http('POST', '/sse')],
  enqueues: [],
} as const satisfies StepConfig

const SAFE_HEADERS = ['content-type', 'user-agent', 'accept', 'accept-language', 'content-length', 'x-request-id']

export const handler: Handlers<typeof config> = async ({ request, response }) => {
  const sanitizedHeaders = Object.fromEntries(
    SAFE_HEADERS.filter((h) => request.headers[h] != null).map((h) => [h, request.headers[h]]),
  )
  logger.info('Data received', { headers: sanitizedHeaders })
  response.status(200)
  const sseHeaders = {
    'content-type': 'text/event-stream',
    'cache-control': 'no-cache',
    connection: 'keep-alive',
  }
  response.headers(sseHeaders)
  logger.info('Headers set', { headers: sseHeaders })

  const chunks: string[] = []

  for await (const chunk of request.requestBody.stream) {
    chunks.push(Buffer.from(chunk).toString('utf-8'))
  }

  const body = chunks.join('').trim()
  const parts: Record<string, string> = {}
  for (const [key, value] of new URLSearchParams(body)) {
    parts[key] = value
  }

  const items = generateRandomItems(parts)

  for (const item of items) {
    response.stream.write(`event: item\ndata: ${JSON.stringify(item)}\n\n`)
    await sleep(300 + Math.random() * 700)
  }

  response.stream.write(`event: done\ndata: ${JSON.stringify({ total: items.length })}\n\n`)
  response.close()
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

function generateRandomItems(parts: Record<string, string>) {
  const fields = Object.entries(parts).map(([name, value]) => ({ name, value }))
  const count = 5 + Math.floor(Math.random() * 6)
  const adjectives = ['swift', 'lazy', 'bold', 'calm', 'fierce', 'gentle', 'sharp', 'wild']
  const nouns = ['falcon', 'river', 'mountain', 'crystal', 'thunder', 'shadow', 'ember', 'frost']

  return Array.from({ length: count }, (_, i) => ({
    id: `item-${Date.now()}-${i}`,
    label: `${adjectives[Math.floor(Math.random() * adjectives.length)]} ${nouns[Math.floor(Math.random() * nouns.length)]}`,
    score: Math.round(Math.random() * 100),
    source: fields.length > 0 ? fields[i % fields.length] : null,
  }))
}
